backend/conftest.py

```diff
@@ -44,3 +44,37 @@ def epistolary_event(db, letter, case_study):
     )
 
     return epistolary_event
+
+
+@pytest.fixture()
+def person_unnamed(db):
+    person = Person.objects.create()
+    return person
+
+
+@pytest.fixture()
+def person_single_name(db):
```

Comment:
In order to test the result of `person.__str__()` if a `Person` has no, one or multiple added names, I need to create several instances of Person somehow (or add names progressively).

I am now creating several fixtures, but I can see how this gets out of hand quite quickly. I am sure there is a more convenient/standard way of performing the "arrange" part of `pytest` tests. Is there a convention/example in the lab that I can follow?

Reply:
We don't have any naming conventions, but here are some things I do in cases like this:

First, you can provide fixtures on different scopes. `/backend/conftest.py` is the highest scope for application-wide fixtures. In this case, you could provide these in `/backend/person/conftest.py` (for fixtures in the person app) or even in `/backend/person/tests/test_person_models.py` (i.e. right in the test file). It depends on where you think the fixtures could be reused. In this case, providing them directly in the test file would be appropriate.

Second, you don't _need_  a fixture if you're only using it once. In this case, you could write your test like so:

```py
def test_single_name(db):
    person = Person.objects.create()
    PersonName.objects.create(person=person, value="Bert")
    assert str(person) == "Bert"
```

That's a perfectly fine test. Making a fixture for a single test can help with readability if your test requires elaborate setup, but this one is pretty short. This can get repetitive between tests, however.

Finally, the state of the database is not preserved between tests unless you want it to be. In this case, you should be able to write your tests with fixtures like these:

```python
@pytest.fixture()
def person(db):
    return Person.objects.create()

@pytest.fixture()
def person_name(db, person):
    return PersonName.objects.create(person=person, name="Bert")

@pytest.fixture()
def person_name_2(db, person):
    return PersonName.objects.create(person=person, name="Ernie")

def test_person_no_name(person):
    # assert str(person) == ...

def test_person_single_name(person, person_name):
    # assert str(person) == ...

def test_person_multiple_name(person, person_name, person_name_2):
    # assert str(person) == ...

```